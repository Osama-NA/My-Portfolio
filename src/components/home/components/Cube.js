import { useState, useRef} from 'react';
import { RoundedBox } from '@react-three/drei';
import {useFrame} from '@react-three/fiber';

export const Cube = ({position, color}) => {
    
    const ref = useRef();
    
    const [scaledOut, isScaledOut] = useState(false);

    const scaleOut = () => isScaledOut(!scaledOut);

    useFrame(( ) => {
        ref.current.rotation.z += 0.009;
    });

    return(
        <RoundedBox 
            args={[1.5, 1.5, scaledOut?2.5:1.5]}
            position={position} 
            radius={0.25} 
            onClick={scaleOut}
            ref={ref} 
        >
            <meshPhongMaterial attach="material" color={color?color:'black'} />
        </RoundedBox>
    )
}