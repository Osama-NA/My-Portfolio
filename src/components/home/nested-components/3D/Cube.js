import { RoundedBox } from '@react-three/drei';
import {useState} from 'react';
import {useFrame} from '@react-three/fiber';
import {useRef} from 'react';

export const Cube = ({position, color}) => {
    
    const ref = useRef();

    const [argZ, setArgZ] = useState(1); // Used in the args attribute in RoundedBox as z
    const [scaledOut, isScaledOut] = useState(false);

    const scaleOut = () => {
        if(!scaledOut){
            setArgZ(argZ + 1);
        } else {
            setArgZ(argZ - 1);
        }

        isScaledOut(!scaledOut);
    }

    useFrame(( ) => {
        ref.current.rotation.z += 0.009;
    });

    return(
        <RoundedBox 
            args={[1.5, 1.5, argZ]} 
            position={position} 
            radius={0.25} 
            onClick={scaleOut}
            ref={ref}
        >
            <meshPhongMaterial attach="material" color={color?color:'black'} />
        </RoundedBox>
    )
}