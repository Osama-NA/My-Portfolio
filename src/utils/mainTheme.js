export const mainTheme = {
    themeColors: {
        aquaGreen: '#25FFA3'
    },
    fontColors: {
        gray: '#707070',
        white: '#fff'
    },
    backgroundColors: {
        darkGray: '#232323',
        darkerGray: '#202020',
        lightGray: '#2F2F2F'
    },
    cubesColors: {
        aquaGreen: 'rgb(0, 190, 90)'
    }
}