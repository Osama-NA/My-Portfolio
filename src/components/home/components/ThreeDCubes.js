import {ThreeDCubesContainer} from '../styles/ThreeDCubes.styled';
import { Canvas } from '@react-three/fiber';
import { Cube } from './Cube.js';
import {Suspense} from 'react';

export const ThreeDCubes = ({ cubesColor }) => {
    return (
        <ThreeDCubesContainer data-testid="threed-cubes" >
            <Suspense fallback={null}>
                <Canvas colorManagement camera={{ position: [0.0001, -0.0001, 5.7] }}> {/* canvas camera adjustment using camera position attribute */}

                    {/* LIGHTING */}
                    <pointLight position={[10, 0, 0]} intensity={.7} />
                    <pointLight position={[-10, 0, 0]} intensity={.7} />
                    <pointLight position={[0, 10, 0]} intensity={.7} />
                    <pointLight position={[0, -10, 0]} intensity={.4} />
                    <pointLight position={[0, 0, 10]} intensity={.2} />
                    <pointLight position={[0, 0, -10]} intensity={.2} />
                    <pointLight position={[-10, 10, 0]} intensity={.3} />
                    <pointLight position={[10, -10, 0]} intensity={.3} />
                    <pointLight position={[0, -10, 10]} intensity={.3} />
                    <pointLight position={[0, 10, -10]} intensity={.3} />
                    <ambientLight intensity={.5} />

                    {/* CUBES */}
                    <Cube position={[-2, 2, 0]} color={cubesColor} /><Cube position={[0, 2, 0]} color={cubesColor} /><Cube position={[2, 2, 0]} color={cubesColor} />
                    <Cube position={[-2, 0, 0]} color={cubesColor} /><Cube position={[0, 0, 0]} color={cubesColor} />
                    <Cube position={[-2, -2, 0]} color={cubesColor} />
                </Canvas>
            </Suspense>
        </ThreeDCubesContainer>
    )
}