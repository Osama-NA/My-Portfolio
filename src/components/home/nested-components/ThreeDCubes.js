import {ThreeDCubesContainer} from './ThreeDCubes.styled';
import { Canvas } from '@react-three/fiber';
import { Cube } from './3D/Cube.js';

export const ThreeDCubes = ({cubesColor}) => {
    return (
        <ThreeDCubesContainer>
            <Canvas colorManagement camera={{position:[0.0001, -0.0001, 5.7]}}>
                <pointLight position={[10, 0, 0]} intensity={.7} />
                <pointLight position={[-10, 0, 0]} intensity={.7} />
                <pointLight position={[0, 10, 0]} intensity={.7} />
                <pointLight position={[0, -10, 0]} intensity={.4} />
                <pointLight position={[0, 0, 10]} intensity={.2} />
                <pointLight position={[0, 0, -10]} intensity={.2} />
                <pointLight position={[-10, 10, 0]} intensity={.3} />
                <pointLight position={[10, -10, 0]} intensity={.3} />
                <pointLight position={[0, -10, 10]} intensity={.3} />
                <pointLight position={[0, 10, -10]} intensity={.3} />
                <ambientLight intensity={.5} />

                <Cube position={[-2, 2, 0]} color={cubesColor} /><Cube position={[0, 2, 0]} color={cubesColor} /><Cube position={[2, 2, 0]} color={cubesColor} />
                <Cube position={[-2, 0, 0]} color={cubesColor} /><Cube position={[0, 0, 0]} color={cubesColor} />
                <Cube position={[-2, -2, 0]} color={cubesColor} />

            </Canvas>
        </ThreeDCubesContainer>
    )
}