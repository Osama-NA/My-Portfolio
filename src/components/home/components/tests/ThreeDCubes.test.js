import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components';
import { cleanup } from '@testing-library/react';
import { mainTheme } from '../../../../utils/mainTheme.js';
import { ThreeDCubes } from '../ThreeDCubes.js';
import { Cube } from '../Cube.js';

afterEach(cleanup);

describe("ThreeDCubes component tests", () => {

    const ThreeDCubesComponent = () => {
        return (
            <ThemeProvider theme={mainTheme}>
                <ThreeDCubes cubesColor="black" />
            </ThemeProvider>
        )
    }

    const CubeComponent = () => {
        return (
            <ThemeProvider theme={mainTheme}>
                <ThreeDCubes cubesColor="black">
                    <Cube />
                </ThreeDCubes>
            </ThemeProvider>
        )
    }

    it("ThreeDCubes component renders without crashing", () => {
        let div = document.createElement('div');
        ReactDOM.render(<ThreeDCubesComponent />, div);
    })

    it("ThreeDCubes component matches snapshot", () => {
        const tree = renderer.create(<ThreeDCubesComponent />).toJSON();
        expect(tree).toMatchSnapshot();
    })

    it("Cube component renders without crashing", () => {
        let div = document.createElement('div');
        ReactDOM.render(<CubeComponent />, div);
    })

    it("Cube component matches snapshot", () => {
        const tree = renderer.create(<CubeComponent />).toJSON();
        expect(tree).toMatchSnapshot();
    })
})