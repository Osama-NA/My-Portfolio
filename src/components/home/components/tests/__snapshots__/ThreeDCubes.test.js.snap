// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeDCubes component tests Cube component matches snapshot 1`] = `
<div
  className="sc-bdfBQB gFXFab"
  data-testid="threed-cubes"
>
  <div
    style={
      Object {
        "height": "100%",
        "overflow": "hidden",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <canvas
      style={
        Object {
          "display": "block",
        }
      }
    />
  </div>
</div>
`;

exports[`ThreeDCubes component tests ThreeDCubes component matches snapshot 1`] = `
<div
  className="sc-bdfBQB gFXFab"
  data-testid="threed-cubes"
>
  <div
    style={
      Object {
        "height": "100%",
        "overflow": "hidden",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <canvas
      style={
        Object {
          "display": "block",
        }
      }
    />
  </div>
</div>
`;
